Fix room joining flow

keianhzo opened the pull request:
We are not correctly failing in case of a room join issue ie. invalid credentials. This PR fixes the flow so we correctly fail and halt the process.

vincentfretin commented on src/naf-dialog-adapter.js:
```diff
@@ -983,7 +978,7 @@ export default class DialogAdapter {
 
         if (!this._reconnectionTimeout) {
-          this._reconnectionTimeout = setTimeout(() => this.reconnect(), this._reconnectionDelay);
+          this._reconnectionTimeout = setInterval(() => this.reconnect(), this._reconnectionDelay);
```
Are you sure about this change? This seems weird to use setInterval here. You don't use the updated reconnectionDelay in this case to retry with more delay for each try. You should put back setTimeout here and clearTimeout at https://github.com/mozilla/hubs/blob/ad709d3f367d9267ef54266b076bd731aeb4522a/src/naf-dialog-adapter.js#L959

keianhzo replied:
Good catch @vincentfretin! Actually we were  just failing without retrying. I've updated so the retry mechanism works and we keep retrying until the connection timeout in https://github.com/mozilla/hubs/blob/f180d53b4dda27a6f239a65e6220978099286040/src/hub.js#L633 expires, which is 90secs now.

johnshaughnessy reviewed (commented):
The PR summary seems to explain the `_reconnectionErrorListener` that you added to exit the scene, but I don't fully understand the other changes. Left some comments with questions

keianhzo commented on src/naf-dialog-adapter.js:
```diff
@@ -744,53 +751,44 @@ export default class DialogAdapter extends EventEmitter {
-      if (this._onOccupantsChanged) {
-        this._onOccupantsChanged(this.occupants);
-      }
-    } catch (err) {
```
Any exception during the join, ie. an exception in the "join" request. We were catching that error there and finishing the connect flow, which led to "virtually" joining the room, as `await this._joinRoom();` was succeeding and resolving the promise, while you weren't really joined.

keianhzo commented on src/hub.js:
```diff
@@ -556,28 +556,19 @@ function handleHubChannelJoined(entryManager, hubChannel, messageDispatch, data)
-      let newHostPollInterval = null;
-
-      // When reconnecting, update the server URL if necessary
+      // We don't need to update the server URL when reconnecting, it's always updated every time we join a room
```
Fixed.

keianhzo commented on src/hub.js:
```diff
@@ -556,28 +556,19 @@ function handleHubChannelJoined(entryManager, hubChannel, messageDispatch, data)
-              scene.setAttribute("networked-scene", { serverURL: newServerURL });
-              adapter.serverUrl = newServerURL;
-            }
-          }, 1000);
```
> adapter.serverUrl = newServerURL

That's my understanding too, that line was not doing anything at all.

> If reticulum reassigns the host for the room will people still successfully connect to the new dialog server? (In other words, this code seems to handle changing the dialog server while the client is running... Is that correct and is this handled elsewhere now?)

My understanding is that this code is only handle the case when wee are trying to reconnect so we get the latest server url from the networked scene to update it in the adapter. Right now we are doing that inside the adapter whenever we try to join the room so we should always be getting the most updated version anyway. But I might be missing some context. I'll add @brianpeiris to this to bring some clarity. 

Can the dialog server url change while the client is running and we need to reconnect to the new one?

johnshaughnessy replied:
> Can the dialog server url change while the client is running and we need to reconnect to the new one?

I thought this was possible. I don't know exactly how it manifests but it seems like on `"connectionstatechange"` events for the `send` or `recv` transports, we redo ICE signalling and in some cases could get a new host from reticulum:
https://github.com/mozilla/hubs/blob/70d763026a07266d49be10b155b79b7d3cc2e60a/src/naf-dialog-adapter.js#L716-L725

https://github.com/mozilla/hubs/blob/70d763026a07266d49be10b155b79b7d3cc2e60a/src/naf-dialog-adapter.js#L281-L288

https://github.com/mozilla/hubs/blob/70d763026a07266d49be10b155b79b7d3cc2e60a/src/naf-dialog-adapter.js#L256-L274


Reticulum is contacted at this line:
https://github.com/mozilla/hubs/blob/70d763026a07266d49be10b155b79b7d3cc2e60a/src/naf-dialog-adapter.js#L267

keianhzo replied:
Oh yep, that's the case when the transport is closed for some reason but signaling is still connected, In that case we get the latest host and recreate transports based on that. I thought you meant that there was a case where Reticulum could rollover and you'd need to connect to a new Dialog server while you were still successfully connected.

johnshaughnessy replied:
The comment in the old code logs,
> "Connecting to new Janus server " + newServerURL
so I wanted to understand why that's no longer necessary. You wrote:
> // We don't need to update the server URL when reconnecting, it's always updated every time we join a room
but I thought joining the room wasn't the _only_ time that we'd have to connect to a new ~~Janus~~ dialog server.

> I thought you meant that there was a case where Reticulum could rollover and you'd need to connect to a new Dialog server while you were still successfully connected.

Yes, that's what I thought happened when Reticulum rolled over but maybe I'm mistaken. I don't see any dialog-specific code in the block that handles reconnecting to reticulum: 
https://github.com/mozilla/hubs/blob/e384f00be3d8d5df210490797d5cb8eaf712bbb0/src/hub.js#L1109

Anyway, if you have a solid understanding of this I don't want to block you -- I am reading through this code for the first time to review this PR.

keianhzo commented on src/naf-dialog-adapter.js:
```diff
@@ -287,20 +286,28 @@ export default class DialogAdapter extends EventEmitter {
+          resolve();
+        } catch (err) {
+          if (isInitial) {
+            this.reconnect();
```
I think we want to try to reconnect until the timer set here expires or we reach the maximum number of retries (which we have never used)
https://github.com/mozilla/hubs/blob/f180d53b4dda27a6f239a65e6220978099286040/src/hub.js#L633

The way it works right now after the PR is that we keep on retrying for 90 secs and then we show the error page.

johnshaughnessy replied:
It makes sense that we want to reconnect, but shouldn't we still resolve the promise? I don't know the details of how javascript handles this or if I'm using the correct vocabulary, but won't there some execution state "parked" at `await new Promise(...)` indefinitely? 

More simply, a code snippet like this seems suspect:

```
await new Promise((resolve, reject) => {
  console.log("Do some work but don't resolve");
});
```

keianhzo replied:
Humm, good catch. I guess nothing happens and at some point the Promise object gets garbage collected but I've been doing some more refactoring and the re-connection listeners are not required anymore now that we are handling the server url during room join so I've gone ahead and removed that code and handled the promise correctly.

keianhzo commented:
@johnshaughnessy Looking at this today, I've realized that if signaling gets disconnected randomly, Protoo has an automatic retry resort. In that case when the WSS is connected again we need to make sure that we are joining the room again otherwise only signaling gets connected but we don't reconnect at the WebRTC level. That's probably the reason why joining was handled inside the open event.

Also every time I look at that join room retry code it makes less and less sense: if signaling is connected chances that room join flow fails are low and if it does why would we want to retry and not just show the error page and simplify that flow? I've also removed the retry, I think it covers a very rare case and makes the connection flow more confusing. Let me know your thoughts.

Also I've updated so we correctly handle the case when signaling is closed (after retrying so in case it's a network blip we will reconnect and persist the cut and if it's permanent we just show the connection error page.

I've also removed janus adapter traces, I don't think that code is used anymore.

johnshaughnessy commented on src/hub.js:
```diff
@@ -1353,7 +1314,7 @@ document.addEventListener("DOMContentLoaded", async () => {
         // HUGE HACK Safari does not like it if the first peer seen does not immediately
         // send audio over its media stream. Otherwise, the stream doesn't work and stays
-        // silent. (Though subsequent peers work fine.) This only affects naf janus adapter
+        // silent. (Though subsequent peers work fine.) This only affects naf dialog adapter
```
You removed the hack so should also remove this comment (and presumably the usage of `track`, `oscillator`, and `stream`)

johnshaughnessy commented on src/naf-dialog-adapter.js:
```diff
@@ -744,53 +739,44 @@ export default class DialogAdapter extends EventEmitter {
-      this.emitRTCEvent("error", "Adapter", () => `Join room failed: ${error}`);
-      error("_joinRoom() failed:%o", err);
+    this._connectSuccess(this._clientId);
+    this._initialAudioConsumerPromise = Promise.all(audioConsumerPromises);
```
I know this is not new code, but want to make a note of it anyway -- It is worrying to me that we wait on connecting to all participants. If a peer is about to disconnect when this happens, do we recover?

johnshaughnessy reviewed (changes requested):
I don't know whether the room rejoin flow is frequently used -- it's a little scary to me to remove that retry logic, but on the other hand I'm not against simplifying if our intention is to gradually make the whole thing more robust. We should just be vigilant about looking for bug reports of failing to join the room in the weeks following this deploy.

Only changes I would like to see before release is to completely remove the safari webrtc hack if you're going to remove it.

brianpeiris reviewed (commented):
It's also surprising to me that we'd want to remove this retry logic. Is it actually sufficient to rely solely on protoo's retry mechanism? Doesn't that only account for signalling disconnects, and not dialog/webrtc disconnects? Also the previous retry code also handled changing the server URL in case a dialog server went down completely, isn't that logic missing now? You've also removed the 90 second connection timeout. We put that in place specifically to handle cases where we would have the user wait on a loading screen forever if something had gone wrong with the connection flow. Are we sure that the "close" event will always fire if something goes wrong, or that `connect()` will throw an error in every case?

I think if we were to release this as it is, we would:

1. Want to make sure to test how it behaves with various disconnect scenarios. For example, a disconnect before signalling, a disconnect after signalling but before webrtc connections are established, and a disconnect after all connections are established. I would expect some retry mechanism to activate in all of those cases.
2. Test the case where you have two dialog servers and one of them goes down. Does the code somehow still switch server URLs and reconnect successfully?
3. Test to make sure that we never leave the user waiting on a loading screen that will never complete. I don't think we fully understood why this was happening, but it would be good to have this fallback anyway. We have definitely seen users just wait at a loading screen for upwards of 15 minutes because they really wanted to get into an event, and we don't want to mislead them into thinking that the connection will eventually establish itself, so a 90 second timeout seemed more reasonable.
4. Add telemetry to ensure that we won't have an increase in number of cases where we would show users the "connect error" screen, effectively ending the session, where previously we might have attempted to reconnect instead.

keianhzo commented:
@brianpeiris Thanks for the review. For some context here's how are currently handling the adapter connection/disconnection through Protoo signaling events:

- Regular flow: "open" -> join room (transports creation)
- Failure before room creation: "failed" (retries) -> "closed" -> connection error page.
- Failure after room creation: "opened" -> "disconnected" (transport destruction) -> "failed" (retries) -> "closed" -> connection error page.
- Recovers from failure: "opened" -> "disconnected" (transport destruction) -> "failed" (retries) -> "opened" -> join room (transports creation)

So unless the Protoo signaling WSS hasn't been opened yet, the Protoo state always transitions to "disconnected" and then calls **failed** for every time it tries to recover from failure. If it can't recover, it transitions to **closed**, otherwise it transitions to **opened**. So we always should either close the connection after retries or reconnect if we manage to recover. As far as I understand Protoo should always transition to "closed" in case of any persisting error and I haven't been able to reproduce any other case where it doesn't but I think it won't do any harm to leave the retry safeguard. I've added a commit to restore it (https://github.com/mozilla/hubs/pull/3935/commits/0c268fdc8f77b4e3964fd43170fe0ab4cc1c81c1).

In case of a Protoo failure, we are currently using the default retry [configuration](https://protoo.versatica.com/#websockettransport35) which takes quite a long to fail in case the issue persists, I've updated to only retry 3 times, I think it's probably more reasonable (https://github.com/mozilla/hubs/pull/3935/commits/0d2817e3170192cd7b631bbf1fa7c9c37b224cdf)

The WebRTC transport disconnection is handled automatically by the WebRTC stack (it retries automatically in a similar fashion as Protoo does) and for the case where it transitions to a **failed** state we handle it through Mediasoup's transport **connectionstatechange** event and restart ICE or recreate the transport depending on the transport state. We are currently getting the Dialog host again in that case and I talked about this with John and I think it's a better idea to use the same host as it was used during the adapter connect, I've updated that (https://github.com/mozilla/hubs/pull/3935/commits/9dbc99baffa55dc5f7efa3bedce1d86e307a3ab5).

brianpeiris commented on src/naf-dialog-adapter.js:
```diff
@@ -292,26 +288,40 @@ export default class DialogAdapter extends EventEmitter {
+    this._protoo = new protooClient.Peer(protooTransport);
 
     this._protoo.on("disconnected", () => {
       this.emitRTCEvent("info", "Signaling", () => `Diconnected`);
```
Spotted a typo.
```suggestion
      this.emitRTCEvent("info", "Signaling", () => `Disconnected`);
```

keianhzo replied:
Fixed

brianpeiris reviewed (commented):
Thanks for explaining the retry logic. That makes me feel more confident about the changes, so I agree these changes put us in a better place. Keeping that 90 second fallback is good too.

I see how caching the dialog params makes sense during the ice restarts, but I'm still not sure if we're handling dialog server disconnects correctly when you have two or more dialog server, and one of the servers dies. We use `getHost` when signalling has succeeded, but shouldn't we also check for a new host when we receive a "failed" signalling event?

keianhzo commented:
@brianpeiris I've updated and added support for server rollovers as we talked. Now in case of a connection failure we check if the server has changed and reconnect using the new URL. I've added a watchdog for the re-connections after joining and both the joining and re-connection watchdog now wait 30 secs, 90 felt a little bit too long.

brianpeiris reviewed (approved):
LGTM!
